File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    private void sol(int h, TreeNode root, Map<Integer, Boolean> m, List<Integer> ans) {
        if (root == null) {
            return;
        }
        if (!m.getOrDefault(h, false)) {
            m.put(h, true);
            ans.add(root.val);
        }
        sol(h + 1, root.right, m, ans);
        sol(h + 1, root.left, m, ans);
    }

    public List<Integer> rightSideView(TreeNode root) {
        Map<Integer, Boolean> m = new HashMap<>();
        List<Integer> ans = new ArrayList<>();
        sol(1, root, m, ans);
        return ans;
    }
}

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void sol(int h,TreeNode* root,map<int,int>&m,vector<int>&ans){
        if(!root){return ;}
        if(m[h]==0){
            m[h]=1;
            ans.push_back(root->val);
        }
        sol(h+1,root->right,m,ans);
        sol(h+1,root->left,m,ans);
    }
    vector<int> rightSideView(TreeNode* root) {
        map<int,int>m;
        vector<int>ans;
        sol(1,root,m,ans);
        return ans;
        
    }
};